### DISCUSSION TITLE
Oqtane.org migrated from DNN to Oqtane

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#1964] [General]
Most people are familiar with the old proverb “the shoemaker's children often go barefoot”… which basically refers to someone who does not have time to help themselves because they are too busy helping others. Well, I finally got around to migrating the https://www.oqtane.org website from DNN to Oqtane. The new site has a modern design, is fully responsive, and contains a lot more useful content than the old site. It uses the Blog module and it also uses the Url Mapping feature to ensure the urls from the old site are mapped to the new site. And the site is blazingly fast... and not only for end users... the administrative perspective feels more like a desktop application than a web application (I will have to create a video to demonstrate at some point).

**Comment by CebiConRepo** [comment]
@sbwalker 

my compliment! It's really fun on the portal, super fast... **Great**...

I've been working with DNN for many years (e.g. www.mypackage.de) and have now programmed > 50 modules for DNN. I've often tried to optimize DNN, resources, IIS, etc., but DNN just can't match the speed of Oqtane.

One of my projects, Oqtane on Raspberry Pi in Hyper-V (Debian) event. with Docker (_but didn't work until now_).

**Comment by Raceeend** [comment]
Unfortunately I get an error when I want to visit the https://www.oqtane.org.
'An error has occurred. This application may no longer respond until reloaded. reload'
What do i wrong?

**Reply by sbwalker** [reply]
are you able to provide some more information ie. what browser you are using, etc..? I only see 1 item in the event log today and it is related to a failed login for a user - but this should not cause an exception. Note that none of the user accounts from the old oqtane.org site were migrated to the new site - and user registration is not currently enabled on the new site.

**Reply by Raceeend** [reply]
Running on Window 10 Enterprise version (10.0.18363), language nl-NL
The problem is with all browsers installed (Chrome Versie 97.0.4692.99, Firefox 96.0.2 (32-bits) or Edge Versie 97.0.1072.69) on this windows machine.

The error is; the page won't load -> on the message 'An error has occurred. This application may no longer respond until reloaded. reload'.
Tested at 2022-01-26T13:45:00Z (14:45 local time = CET)

But on the iMac with Chrome Versie 97.0.4692.99 or Safari Versie 13.1.2 it works like charm. (iMac is also in Dutch ->language nl-NL)
It works also on a chromebook (nl-NL)

**Reply by sbwalker** [reply]
can you try clearing your cookies on the Windows 10 machine? I am wondering if that specific machine may have some cookies for the old oqtane.org site which is conflicting with the new site. This is just a wild guess as I am unable to reproduce the problem on any of my machines and based on the logs I can see that the site is successfully serving content to many visitors - and there are no exceptions in the Event Log.

**Reply by Raceeend** [reply]
Done in Firefox no success
Done in Edge no success (all browser history cleared)

**Reply by sbwalker** [reply]
if you look in your Firefox browser Storage/Cookies do you see any items that were created when you visited the site?

**Reply by Raceeend** [reply]
I think that it is a websocket problem...
Attached 2 screenshots form Firefox
![2022-01-26 16_02_40-Window](https://user-images.githubusercontent.com/669420/151188749-c749c895-5081-4bfa-9251-04ef0b9a01c3.jpg)
![2022-01-26 16_03_11-Window](https://user-images.githubusercontent.com/669420/151188765-300c1836-4bf8-4032-8312-a2ccb734ab1c.jpg)

**Reply by sbwalker** [reply]
Very strange indeed! My screenshot from Firefox looks like:

![image](https://user-images.githubusercontent.com/4840590/151194067-76d6ddc0-deae-41a8-b74b-f58526027443.png)

**Reply by sbwalker** [reply]
What information is in your console? Here is a screenshot of mine on a page reload:

![image](https://user-images.githubusercontent.com/4840590/151195359-b2a48831-d9a7-44f5-a923-30b002491a99.png)

**Reply by Raceeend** [reply]
It only happens on the company's PCs.
Most likely the websocket is blocked by company policy

So far I have only run oqtane as a server and that without any problems
![2022-01-26 16_42_19-Clipboard](https://user-images.githubusercontent.com/669420/151196570-6419c34e-813a-450d-ba91-8857afb4ca20.jpg)
.

**Reply by Raceeend** [reply]
It looks like een problem in the security of the company's (personal) laptops.
On a workstation in the company's domain, it works like charm.

The problem is our problem ;-)

**Reply by Raceeend** [reply]
FYI: Indeed the websocket were blocked in our proxy. A rule has been added that websockets are allowed for the oqtane sites.

@Shaun
You did a fantastic job. The site is very nice and running smooth.  Chapeau ;-)

**Comment by lanthonyneville** [comment]
Fantastic work Shaun ... that dogfood should taste pretty good :)

Would you be willing to share the theme as OS on [oqtane.net](https://www.oqtane.net/)? I am a competent .NET developer but very junior with HTML/CSS, and almost zero JS knowledge (I guess that makes me the target consumer for Blazor, right?) .... and this is the first theme I have seen that doesn't look like a Blazor sample app.

Thanks again for your work.

**Comment by Sharafudeen** [comment]
Wow.. simply awesome and really great job..!!

In a very short time period.. Oqtane's features are great, promising & would like to commit for world-class enterprise systems..
Hope it will be dominated world market upcoming decades!

@sbwalker you are game changer.. :)

**Comment by sbwalker** [comment]
Thank you all for your kind words... I promise to provide some more information about this theme shortly. However I am still struggling with one design issue which I am hoping someone may be able to clarify for me. The site validates fine on the W3C Validator and behaves properly in all browsers I have tried, at all screen resolutions. However when I generate a screen shot using https://restpack.io/screenshot and choose the PNG and Full Page options it produces a result which is very strange:

https://restpack.io/cache/screenshot/e9a68792f433802e91caac336b84e7dab52e33fd2badbd06856568a47b30c4f2

This is an issue because the Built On Blazor! showcase site ( https://builtonblazor.net/ ) uses this service/method to generate screen shots and the current screen shot for oqtane.org does not demonstrate the capabilities of the framework in the best possible way.

Any suggestions would be greatly appreciated...

**Reply by Sharafudeen** [reply]
I've tried in Microsoft Edge and it is takes screenshots without any issues.

**Web Capture:**
![image](https://user-images.githubusercontent.com/42769344/151512854-45a7ecf1-dadd-4322-829d-f7e823209267.png)

**Screenshot Oqtane:**
![image](https://user-images.githubusercontent.com/42769344/151512997-473ee0eb-57ac-48c3-94a9-6572fcf7daf1.png)

**Reply by sbwalker** [reply]
Yes it is specifically Restpack.io's service which is producing strange results - but the service seems to work fine for the other sites in the BuiltOnBlazor showcase - so there must be quirky design issue in the Oqtane.org theme.

**Reply by sbwalker** [reply]
The issue has been resolved:

https://restpack.io/cache/screenshot/2b78a16eb27b3a8bae14ed1fa1797a95719664db674192b49fc7e0bcba6e2696

**Comment by sbwalker** [comment]
So.... because Oqtane utilizes Blazor and Bootstrap, it is not difficult to create an Oqtane theme. Basically you can go online and search for any free or commercial Bootstrap template which meets your requirements. Usually these templates are provided in the form of HTML, CSS, JavaScript, and possibly some stock images. In my case I purchased a Bootstrap template - so I cannot share it as an open source project.

In order to transform the Bootstrap template into an Oqtane theme, the simplest approach is to scaffold your Oqtane theme by going to the Theme Management area of the Admin Dashboard and choose the Create Theme button. This will allow you to provide some basic info about your theme and then generate all of the code automatically. Once the code is generated you can open the solution file in Visual Studio to develop your template. 

You can copy all of the content within the body tags of your Bootstrap template into your Default.razor theme file. All of the head references to CSS files in the Bootstrap template will need to be defined as Resources in your Default.Razor theme file. Make sure you include the UserProfile, Login, and ControlPanel components and an Admin Pane reference in your theme markup. Use ThemePath() to set the path to any images. 

The most challenging aspect of migrating a Bootstrap template is dealing with any JavaScript references as Blazor requires you to use JS Interop. If there are elements of JavaScript in the Bootstrap template which are not essential I would recommend removing them (ie. I removed references to AOS - animate on scroll library because there were not necessary for my needs). I also included content directly in the Oqtane theme file as in most cases this marketing content does not change often - and it is trivial to deploy a new version of the theme if I need to change any content. 

Here is a slimmed down version of the razor file:

```
@namespace Oqtane.Org
@inherits ThemeBase

  <header id="header" class="fixed-top d-flex align-items-center">
    <div class="container d-flex align-items-center">
      <a href="@(PageState.Alias.Path)/" class="logo me-auto"><img src="@ThemePath()assets/img/logo.png" alt="Oqtane - Modular Application Framework for Blazor"></a>

      <nav id="navbar" class="navbar order-last order-lg-0">
        <ul>
          <li><a class="nav-link" href="@(PageState.Alias.Path)/#home">Home</a></li>
          <li><a class="nav-link" href="@(PageState.Alias.Path)/#about">About</a></li>
          <li><a class="nav-link" href="@(PageState.Alias.Path)/#features">Features</a></li>
          <li><a class="nav-link" href="@(PageState.Alias.Path)/#showcase">Showcase</a></li>
          <li><a class="nav-link" href="@(PageState.Alias.Path)/#license">License</a></li>
          <li><a class="nav-link" href="@(PageState.Alias.Path)/blog">Blog</a></li>
        </ul>
        <i class="bi bi-list mobile-nav-toggle"></i>
      </nav>
      <div class="mx-3 controls-group"><UserProfile /> <Login /> <ControlPanel /></div>
    </div>
  </header>

  <main id="main">
    <section id="default" class="default">
        <Pane Name="Admin"></Pane>
    </section>
  </main>

  <footer id="footer">
    <div class="footer-top">
      <div class="container">
        <div class="row">
          <div class="col-lg-3 col-md-6 footer-contact">
            <p><img src="@ThemePath()assets/img/logo_white.png" alt="Oqtane - Modular Application Framework for Blazor"></p>
          </div>
          <div class="col-lg-2 col-md-6 footer-links">
            <h4>Useful Links</h4>
            <ul>
              <li><i class="bx bx-chevron-right"></i> <a href="https://blazorhelpwebsite.com/ViewBlogPost/1" target="_blank">Installing on Azure</a></li>
              <li><i class="bx bx-chevron-right"></i> <a href="https://docs.oqtane.org" target="_blank">Documentation</a></li>
              <li><i class="bx bx-chevron-right"></i> <a href="https://www.nuget.org/profiles/oqtane" target="_blank">Nuget Packages</a></li>
            </ul>
          </div>
          <div class="col-lg-3 col-md-6 footer-links">
            <h4>Releases</h4>
            <ul>
              <li><i class="bx bx-chevron-right"></i> <a href="Resources/Blog/PostId/551/announcing-oqtane-30-for-net-6">Oqtane 3.0</a></li>
              <li><i class="bx bx-chevron-right"></i> <a href="Resources/Blog/PostId/549/oqtane-22-upgrades-to-bootstrap-5">Oqtane 2.2</a></li>
            </ul>
          </div>
          <div class="col-lg-4 col-md-6 footer-newsletter">
            <h4>Join Our Community</h4>
            <p>Open source projects are dependent upon the size and passion of their communities</p>
            <a href="https://github.com/oqtane/oqtane.framework/discussions" class="btn btn-danger" target="_blank">Join Now</a>
          </div>
        </div>
      </div>
    </div>
    <div class="container d-md-flex py-4">
      <div class="me-md-auto text-center text-md-start">
        <div class="copyright">
          &copy; Copyright 2019-2022 <strong><span>.NET Foundation</span></strong>. All Rights Reserved.
        </div>
        <div class="credits">
          Powered by <a href="https://www.oqtane.org/">Oqtane</a> on Blazor...
        </div>
      </div>
      <div class="social-links text-center text-md-end pt-3 pt-md-0">
        <a href="https://twitter.com/oqtane" class="twitter" target="_blank"><i class="bx bxl-twitter"></i></a>
        <a href="https://www.linkedin.com/in/shaunbrucewalker/" class="linkedin" target="_blank"><i class="bx bxl-linkedin"></i></a>
      </div>
    </div>
  </footer>

  <a href="#" class="back-to-top d-flex align-items-center justify-content-center"><i class="bi bi-arrow-up-short"></i></a>

@code {
    public override string Name => "Default Page";

    public override string Panes => PaneNames.Admin;

    public override List<Resource> Resources => new List<Resource>()
    {
        new Resource { ResourceType = ResourceType.Stylesheet, Url = "https://fonts.googleapis.com/css?family=Open+Sans:300,300i,400,400i,600,600i,700,700i%7CRaleway:300,300i,400,400i,500,500i,600,600i,700,700i%7CPoppins:300,300i,400,400i,500,500i,600,600i,700,700i" },
        new Resource { ResourceType = ResourceType.Stylesheet, Url = ThemePath() + "assets/vendor/bootstrap/css/bootstrap.min.css" },
        new Resource { ResourceType = ResourceType.Stylesheet, Url = ThemePath() + "assets/vendor/bootstrap-icons/bootstrap-icons.css" },
        new Resource { ResourceType = ResourceType.Stylesheet, Url = ThemePath() + "assets/vendor/boxicons/css/boxicons.min.css" },
        new Resource { ResourceType = ResourceType.Stylesheet, Url = ThemePath() + "assets/css/style.css" },
        new Resource { ResourceType = ResourceType.Script, Url = ThemePath() + "assets/vendor/bootstrap/js/bootstrap.bundle.min.js" },
        new Resource { ResourceType = ResourceType.Script, Url = ThemePath() + "assets/js/interop.js" }
	};

    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        await base.OnAfterRenderAsync(firstRender);

        if (firstRender)
        {
            var interop = new Interop(JSRuntime);
            await interop.Load();                
        }
    }
}

```

So @lanthonyneville you really dont need to be strong with web design to create an Oqtane theme - simply leverage some of the amazing Bootstrap templates which are available.
